### PR TITLE
Check if the parent folder is updatable when moving

### PR DESCRIPTION
**Pull request by juliushaertl**
When moving folders we should also check if the parent folder is actually updatable. Otherwise users could mess with the folder structure.

Try moving FolderA into FolderB with the following ACL applied:
- Groupfolder/  READ 
- Groupfolder/FolderA/  READ WRITE UPDATE DELETE
- Groupfolder/FolderB/  READ WRITE UPDATE DELETE

### CODE REVIEW

File: lib/ACL/ACLStorageWrapper.php [hunk 1]
```diff
@@ -88,7 +88,9 @@ public function rename($path1, $path2) {
 			}
 		}
 		$permissions = $this->file_exists($path2) ? Constants::PERMISSION_UPDATE : Constants::PERMISSION_CREATE;
-		return $this->checkPermissions($path1, Constants::PERMISSION_UPDATE & Constants::PERMISSION_READ) and
+		$checkParentPermissions = !$this->is_dir($path1) or $this->checkPermissions(dirname($path1), Constants::PERMISSION_UPDATE);
+		return $checkParentPermissions and
```
Comment (nickvergessen) [inline lib/ACL/ACLStorageWrapper.php]:
```suggestion
		return $checkParentPermissions &&
```

File: lib/ACL/ACLStorageWrapper.php [hunk 2]
```diff
@@ -88,7 +88,9 @@ public function rename($path1, $path2) {
...
+			$this->checkPermissions($path1, Constants::PERMISSION_UPDATE & Constants::PERMISSION_READ) and
```
Comment (nickvergessen) [inline lib/ACL/ACLStorageWrapper.php]:
```suggestion
			$this->checkPermissions($path1, Constants::PERMISSION_UPDATE & Constants::PERMISSION_READ) &&
```

File: lib/ACL/ACLStorageWrapper.php [hunk 3]
```diff
@@ -88,7 +88,9 @@ public function rename($path1, $path2) {
...
 			$this->checkPermissions($path2, $permissions) and
```
Comment (nickvergessen) [inline lib/ACL/ACLStorageWrapper.php]:
```suggestion
			$this->checkPermissions($path2, $permissions) &&
```

**Comment by juliushaertl** [comment]
> They are not equal, and in fact and means the options are all tested, so parent::rename() is actually called.

Ah thanks, I wasn't aware of that :confounded: I made sure that `&&` is now used all over the place, since other operations were also affected by this.

File: lib/ACL/ACLStorageWrapper.php [hunk 4]
```diff
@@ -88,7 +88,9 @@ public function rename($path1, $path2) {
 			}
 		}
 		$permissions = $this->file_exists($path2) ? Constants::PERMISSION_UPDATE : Constants::PERMISSION_CREATE;
-		return $this->checkPermissions($path1, Constants::PERMISSION_UPDATE & Constants::PERMISSION_READ) and
+		$checkParentPermissions = !$this->is_dir($path1) or $this->checkPermissions(dirname($path1), Constants::PERMISSION_UPDATE);
```
Comment (icewind1991) [inline lib/ACL/ACLStorageWrapper.php]: Why only do this check for folders?

**Comment by skjnldsv** [comment]
@nickvergessen @icewind1991 can any of you help on this pr?
This is high priority and Julius is in vacations :pray:

**Comment by skjnldsv** [comment]
Thanks @icewind1991 :heart:

**Pull request merged**